WIP: pluggable upload destinations

tedder opened the pull request:
Since uploads.im was going away I wanted to store my images on S3. I'm sure
that's not what most people want to do, so I abstracted it a little. This
strategy would be a good way to go for sending messages (eg twilio, gvoice).

Marking as a WIP so you all can see it, and because it's a WIP. I'm still testing, and I need to add options to the config panel.

tedder committed:
WIP: pluggable upload destinations

Since uploads.im was going away I wanted to store my images on S3. I'm sure
that's not what most people want to do, so I abstracted it a little. This
strategy would be a good way to go for sending messages (eg twilio, gvoice).

Marking as a WIP so you all can see it, and because it's a WIP. I'm still testing, and I need to add options to the config panel.

tedder committed:
add boto3 dependency, fix plugin url

taxilian commented:
@shadycuz I don't have time to review this right now, any chance you do?

shadycuz commented:
Looks like a really nice upload class. Lots of flexibility for the future. @tedder Let me test this, and fleshout what the UI should look like in the settings tab. I'll push any changes to your branch. 

Thanks,
Levi

tedder commented:
Thanks. I'm still iterating and testing, I'll push my fixes now.

shadycuz commented:
@tedder oh yes please do what you do ;)

I am torn about the new function for sending a text with image. To me the body of the text stays the same and the image is optional. Which is why I used the one function with optional url. 

Your new function is cleaner but maybe unnecessary? Maybe it would help to move the message formatting to it's own function and pass the body into the function?

For example we allow messages to go to more than one number, so maybe we have a function that produces a dict where the keys are numbers and the values are the formatted txt message.
```json
{
   "+1-555-555-5555" : "This would be the text sent"
}
```

Doesn't even look like multiple numbers is working, I probably broke it 👎 

https://github.com/taxilian/OctoPrint-Twilio/blob/5e1dd4fe2a7f20612512f8631b3c6d14ffdd6488/octoprint_smsnotifier/__init__.py#L100-L109

I'm just thinking out loud.

tedder commented:
What I was trying to do with the second send() function was to centralize the "try with image, then try without image" stuff. It should work for all cases now, right?

I haven't used the multiple phone stuff. Once I'm sure it's actually working I'll look at that. Message formatting elsewhere is a decent idea.

shadycuz commented:
I thought it worked for all cases before?

tedder committed:
incremental fixes

- make object/super work better across py2/py3
- set up logger better
- fix config names in s3
- pass a 'suggested filename'
- switch from presigned URL to one created by config
- fix send_text call

tedder committed:
fix datetime

tedder committed:
remove my personal config settings from this

tedder committed:
wrap the 'send' so we can send to multiple phone numbers without nesting

tedder commented:
@shadycuz I fixed the 'multiple phone numbers' thing.

I've been using this branch constantly since putting in the PR, works well with S3.

shadycuz commented on octoprint_smsnotifier/.flake8:
```diff
@@ -0,0 +1,5 @@
+max-line-length= 120
+
+ignore=
+ E127, E128 # bad indentation for continuation lines; would result in really deep indentation, which is the whole reason for doing continuation lines.
```
This .flake8 is specific to your IDE? Maybe we should remove it?

tedder replied:
vim is my IDE; I have it to make sure I keep to your indentation style and such. So it's specific to this repo.

shadycuz commented on octoprint_smsnotifier/__init__.py:
```diff
@@ -24,6 +151,11 @@ def get_settings_defaults(self):
             auth_token="",
             printer_name="",
+            #upload_provider="cloudinary",
+            upload_provider="aws_s3",
```
You have this hardcoded correct? I will need to adjust the setting plugin page to account for this. I'll try to have it done by next weekend.

tedder replied:
Saw I failed to respond here. The hardcoding is an artifact- I specify it in `config.yaml`.

shadycuz replied:
Config.yml? I don't see it in the repo? I'm guessing it wont matter once I rebuild the config page. As you can see its been a couple weekends and I haven't taken a look at it yet.

tedder replied:
Octoprint's config.yaml:

```
plugins:
...
  smsnotifier:
    _config_version: 1
    account_sid: xxx
    auth_token: xxx
    body: '{printer_name} job complete after {elapsed_time}'
    enabled: true
    from_number: '1234'
    printer_name: e10
    recipient_number: '1234'
    upload_provider: aws_s3
    s3_bucket: dyn.tedder.me
    s3_key_prefix: 3d/automated/
    s3_url: https://dyn.tedder.me
    send_image: true
```

So twilio tokens are in the config, AWS credentials are in `~/.aws/credentials`.

It's been a while for me too :)

shadycuz replied:
oh you mean you can set plugin variables from the config.yml and not have to do it in the GUI?

tedder replied:
Exactly. Handy because I haven't modified the GUI to support all the necessary vars.

shadycuz replied:
ahh I was wondering how that would work. Well I think I can merge this into a test branch, and modify it back so cloudinary is the default and we can leave the other API backends as advanced use and make some documentation on how to set it up. We can leave the configuration page to another PR?

tedder replied:
Yeah, I like burying the switches. It makes me happy and yet it works by default.

I'll change the cloudinary thing and push it here, which'll be easier for you.

shadycuz replied:
=) thanks

tedder committed:
need to collect and return success from here

tedder committed:
cloudinary by default; comment about config.yaml

tedder commented:
I'm going to be setting up a few Pis in the next week and want this feature- can you merge or should I just fork the plugin?

shadycuz commented:
We want your continued development in here. Let me get some coffee and look it over again and then I'll merge.

shadycuz commented:
@tedder I'll get this merged this weekend.

taxilian commented:
@shadycuz can't help notice that you indicated you were going to merge this but didn't; is it ready to be merged?

shadycuz commented:
@taxilian I will try it again soon.
